File: apps/api/app/Constants/ErrorCodes.php
<?php

namespace App\Constants;

class ErrorCodes
{
    public const INVALID_PHONE = 'INVALID_PHONE';
    public const OTP_EXPIRED = 'OTP_EXPIRED';
    public const OTP_INVALID = 'OTP_INVALID';
    public const OTP_RATE_LIMIT = 'OTP_RATE_LIMIT';
    public const OTP_MAX_ATTEMPTS = 'OTP_MAX_ATTEMPTS';
    public const UNAUTHORIZED = 'UNAUTHORIZED';
    public const FORBIDDEN = 'FORBIDDEN';
    public const NOT_FOUND = 'NOT_FOUND';
    public const VALIDATION_ERROR = 'VALIDATION_ERROR';
    public const DUPLICATE_NID = 'DUPLICATE_NID';
    public const DUPLICATE_PLATE = 'DUPLICATE_PLATE';
    public const NO_DRIVERS = 'NO_DRIVERS';
    public const OUT_OF_ZONE = 'OUT_OF_ZONE';
    public const INVALID_TRANSITION = 'INVALID_TRANSITION';
    public const PIN_MISMATCH = 'PIN_MISMATCH';
    public const ALREADY_PAID = 'ALREADY_PAID';
    public const SOS_NOT_ALLOWED = 'SOS_NOT_ALLOWED';
    public const GUARDIAN_CAP = 'GUARDIAN_CAP';
    public const DUPLICATE_GUARDIAN = 'DUPLICATE_GUARDIAN';
    public const DEBT_CAP = 'DEBT_CAP';
    public const GRACE_OVER = 'GRACE_OVER';
    public const SMS_FAILED = 'SMS_FAILED';
    public const ALREADY_RATED = 'ALREADY_RATED';
    public const BLOCKED = 'BLOCKED';
}

File: packages/mobile_core/lib/core/network/error_codes.dart
/// Must match packages/shared-constants and apps/api/app/Constants/ErrorCodes.php
abstract final class ErrorCodes {
  static const invalidPhone = 'INVALID_PHONE';
  static const otpExpired = 'OTP_EXPIRED';
  static const otpInvalid = 'OTP_INVALID';
  static const otpRateLimit = 'OTP_RATE_LIMIT';
  static const otpMaxAttempts = 'OTP_MAX_ATTEMPTS';
  static const unauthorized = 'UNAUTHORIZED';
  static const forbidden = 'FORBIDDEN';
  static const notFound = 'NOT_FOUND';
  static const validationError = 'VALIDATION_ERROR';
  static const duplicateNid = 'DUPLICATE_NID';
  static const duplicatePlate = 'DUPLICATE_PLATE';
  static const noDrivers = 'NO_DRIVERS';
  static const outOfZone = 'OUT_OF_ZONE';
  static const invalidTransition = 'INVALID_TRANSITION';
  static const pinMismatch = 'PIN_MISMATCH';
  static const alreadyPaid = 'ALREADY_PAID';
  static const sosNotAllowed = 'SOS_NOT_ALLOWED';
  static const guardianCap = 'GUARDIAN_CAP';
  static const duplicateGuardian = 'DUPLICATE_GUARDIAN';
  static const debtCap = 'DEBT_CAP';
  static const graceOver = 'GRACE_OVER';
  static const smsFailed = 'SMS_FAILED';
  static const alreadyRated = 'ALREADY_RATED';
  static const blocked = 'BLOCKED';

  // Legacy aliases used in UI copy paths
  static const badPhone = invalidPhone;
  static const rateLimit = otpRateLimit;
  static const smsDown = smsFailed;
  static const duplicateNidPlate = duplicateNid;
}

/// Must match packages/shared-types/src/socket-events.ts and SocketEvents.php
abstract final class SocketEvents {
  static const driverLocationUpdate = 'driver:location:update';
  static const driverStatusUpdate = 'driver:status:update';
  static const rideAccept = 'ride:accept';
  static const rideDecline = 'ride:decline';
  static const rideArrived = 'ride:arrived';
  static const rideCashConfirm = 'ride:cash:confirm';
  static const rideComplete = 'ride:complete';
  static const rideCancel = 'ride:cancel';
  static const serverRideDispatched = 'server:ride:dispatched';
  static const serverRideCancelled = 'server:ride:cancelled';
  static const serverRideTimeout = 'server:ride:timeout';
  static const serverRideAccepted = 'server:ride:accepted';
  static const serverDriverLocation = 'server:driver:location';
  static const serverDriverArrived = 'server:driver:arrived';
  static const serverRideStarted = 'server:ride:started';
  static const serverRideCompleted = 'server:ride:completed';
  static const serverDriverCancelled = 'server:driver:cancelled';
  static const adminSosAlert = 'admin:sos:alert';
  static const adminDashboard = 'admin:dashboard';
}

enum RideStatus {
  requested,
  accepted,
  driverArriving,
  driverArrived,
  inProgress,
  completed,
  cancelled,
  noShow,
  noDriverAvailable,
}

enum DriverOnboardingStatus {
  phoneVerified,
  personalDone,
  vehicleDone,
  docsPending,
  pendingReview,
  rejected,
  approved,
  grace,
}

class ApiException implements Exception {
  ApiException({required this.code, required this.message, this.status = 400});
  final String code;
  final String message;
  final int status;

  @override
  String toString() => message;
}
